packages/react-dom/src/client/ReactDOM.js

```diff
@@ -659,14 +659,36 @@ const DOMRenderer = ReactFiberReconciler({
   },
 
   shouldSetTextContent(type: string, props: Props): boolean {
-    return (
-      type === 'textarea' ||
+    if (type === 'textarea') {
+      return true;
+    }
+
+    if (
       typeof props.children === 'string' ||
-      typeof props.children === 'number' ||
-      (typeof props.dangerouslySetInnerHTML === 'object' &&
-        props.dangerouslySetInnerHTML !== null &&
-        typeof props.dangerouslySetInnerHTML.__html === 'string')
-    );
+      typeof props.children === 'number'
+    ) {
+      return true;
+    }
+
+    if (
```

Comment:
Overall: we are reading `props.dangerouslySetInnerHTML` a *lot* of times here. I know old code did too, but this makes reading a bit difficult in all further conditions. Mind extracting it to a variable?